### PR TITLE
DocumentMarkerLineStyle should encode the marker color

### PR DESCRIPTION
**Pull request by pxlcoder**
#### 38b829c238b8b749bd68bf25b5715acb90bde295
<pre>
DocumentMarkerLineStyle should encode the marker color
<a href="https://bugs.webkit.org/show_bug.cgi?id=252669">https://bugs.webkit.org/show_bug.cgi?id=252669</a>
rdar://105729542

Reviewed by Wenson Hsieh.

Currently, document marker colors are determined at the GraphicsContext level.
To support the use of system colors for marker colors, the colors must be
resolved earlier, in the web process.

* Source/WebCore/platform/graphics/GraphicsTypes.h:

Add a `color` member to `DocumentMarkerLineStyle`, representing the marker
color. Drop the `shouldUseDarkAppearance` bit, as the `color` already takes
light/dark mode into account.

* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawDotsForDocumentMarker):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContextCG::drawDotsForDocumentMarker):
(WebCore::grammarColor): Deleted.
(WebCore::colorForMarkerLineStyle): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItemBuffer.cpp:
(WebCore::DisplayList::ItemHandle::destroy):
* Source/WebCore/platform/graphics/displaylists/DisplayListItemType.cpp:
(WebCore::DisplayList::isInlineItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawDotsForDocumentMarker::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:

`DrawDotsForDocumentMarker` can no longer be an inline item, since `Color` can
have out-of-line components.

(WebCore::DisplayList::DrawDotsForDocumentMarker::DrawDotsForDocumentMarker):
(WebCore::DisplayList::DrawDotsForDocumentMarker::style const):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::documentMarkerLineColor const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::grammarColor):
(WebCore::RenderThemeCocoa::documentMarkerLineColor const):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarker):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/260658@main">https://commits.webkit.org/260658@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3ac3845f6cbaac174f414cb3a63fc919b90e74c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108905 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41748 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/451 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118223 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9294 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101205 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114675 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97827 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42758 "Found 4 new test failures: webgl/2.0.y/conformance/glsl/bugs/character-set.html, webgl/2.0.y/conformance/glsl/misc/fragcolor-fragdata-invariant.html, webgl/2.0.y/conformance/misc/expando-loss.html, webgl/2.0.y/conformance2/extensions/promoted-extensions-in-shaders.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96577 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29468 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84505 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10784 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30817 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11538 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7749 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16928 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50413 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13128 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4022 "Built successfully and passed tests") | | | |

### CODE REVIEW

File: Source/WebCore/rendering/TextBoxPainter.cpp [hunk 1]
```diff
@@ -754,30 +755,30 @@ void TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarker(const MarkedText&
 
     auto bounds = calculateDocumentMarkerBounds(makeIterator(), markedText);
 
-    auto lineStyleForMarkedTextType = [&]() -> DocumentMarkerLineStyle {
-        bool shouldUseDarkAppearance = m_renderer.useDarkAppearance();
+    auto lineStyleMode = [&]() -> DocumentMarkerLineStyleMode {
```
Comment (rr-codes) [inline Source/WebCore/rendering/TextBoxPainter.cpp]: is the `() -> DocumentMarkerLineStyleMode` needed?

File: Source/WebCore/rendering/RenderThemeCocoa.mm [hunk 2]
```diff
@@ -242,4 +243,36 @@ static inline FontSelectionValue cssWeightOfSystemFont(CTFontRef font)
...
+{
+#if ENABLE(POST_EDITING_GRAMMAR_CHECKING)
+    static bool useBlueForGrammar = false;
+    static std::once_flag flag;
+    std::call_once(flag, [] {
+        useBlueForGrammar = os_feature_enabled(TextComposer, PostEditing) && os_feature_enabled(TextComposer, PostEditingUseBlueDots);
+    });
+
+    if (useBlueForGrammar)
+        return useDarkMode ? SRGBA<uint8_t> { 40, 145, 255, 217 } : SRGBA<uint8_t> { 0, 122, 255, 191 };
```
Comment (rr-codes) [inline Source/WebCore/rendering/RenderThemeCocoa.mm:256]: maybe these hard-coded colors should be refactored out to `constexpr` constants for better readability?

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.h [hunk 3]
```diff
@@ -951,34 +951,23 @@ std::optional<DrawLinesForText> DrawLinesForText::decode(Decoder& decoder)
 class DrawDotsForDocumentMarker {
 public:
     static constexpr ItemType itemType = ItemType::DrawDotsForDocumentMarker;
-    static constexpr bool isInlineItem = true;
+    static constexpr bool isInlineItem = false;
```
Comment (rr-codes) [inline Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:954]: why was this changed?

Comment (pxlcoder) [reply Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:954]: See ChangeLog:

> `DrawDotsForDocumentMarker` can no longer be an inline item, since `Color` can
have out-of-line components.

File: Source/WebCore/rendering/TextBoxPainter.cpp (hunk 1 above)
Comment (pxlcoder) [reply Source/WebCore/rendering/TextBoxPainter.cpp]: Removed.

File: Source/WebCore/rendering/RenderThemeCocoa.mm (hunk 2 above)
Comment (pxlcoder) [reply Source/WebCore/rendering/RenderThemeCocoa.mm:256]: Some of these will start using system colors.

For the others, I don't think there's a big readability improvement, given that the resolved color is either factored into a named function, or inline with the marker type. I also can't think of names that would make the values any clearer.

Given that this is existing code, and I don't think there's a readability issue here, I will keep this as-is unless you feel strongly.

**Review by whsieh — APPROVED** [review]
